README.md
# pgpx
"Single-backend" PostgreSQL proxy for enabling RDS IAM-based authentication with dynamic passwords on Postgres client apps that do not support dynamic credentials.

## How it works

`pgpx` works as a dumb proxy for PostgreSQL with a built-in ability to intercept the auth requests and substitute the authentication password with the dynamic one, generated by the bash command (usually AWS CLI) provided as argument to the script.

## Usage

Here's the 'official' usage output:

```bash
Usage of ./pgpx:
  -credsCmd string
    	Bash command to retrieve temporary database connection password
  -credsOverride
    	Indicates whether to override the password provided through the client application (default true)
  -ignoreUsername
        By default, the tool matches provided username with the one in credsCmd command to decide on overriding the password.
        Set this flag to ignore username matching. (default false)
  -dbCertPath string
    	Absolute path to the database SSL CA certificate (required)
  -pgHostPort string
    	Hostname and port of the PostgreSQL server, e.g. "localhost:5432" (required)
  -proxyHost string
    	Host where pgpx is accepting connections (default "localhost")
  -proxyPort string
    	Port where pgpx is accepting connections (default "6432")
```

## Real-world example

**1. Prepare AWS CLI command to retrieve the dynamic database password**

The typical AWS CLI command to generate the RDS IAM dynamic credentials looks like the below:

```bash
AWS_PROFILE=${AWS_PROFILE} aws --region ${REGION} \
  rds generate-db-auth-token --hostname ${RDS_HOSTNAME} --port ${RDS_PORT} \
  --username "${IAM_USERNAME}" --output text
```

where 
- `AWS_PROFILE` is a name of the AWS profile
- `REGION` is a target AWS region
- `RDS_HOSTNAME` is a host name of the RDS server
- `RDS_PORT` is a port of the RDS server
- `IAM_USERNAME` is the username of the user for whom the credentials are retrieved

**NOTE.** Make sure that RDS_HOSTNAME and RDS_PORT are directly accessible for `pgpx`, otherwise the proxy will not work.

**2. Export the dynamic password command as an env varable**

Just for simplicity, let's export the dynamic password command as a `DYNAMIC_PASSWORD` env var.

```bash
export DYNAMIC_PASSWORD=(AWS_PROFILE=${AWS_PROFILE} aws --region ${REGION} rds generate-db-auth-token --hostname ${RDS_HOSTNAME} --port ${RDS_PORT} --username "${IAM_USERNAME}" --output text)
```

**3. Keep database SSL/TLS certificates handy**

Make sure you know the path to RDS SSL certificate that will be used to establish a secure connection to the database.

The AWS article [Using SSL/TLS to encrypt a connection to a DB instance](https://docs.aws.amazon.com/AmazonRDS/latest/UserGuide/UsingWithRDS.SSL.html) explains how to establish SSL/TLS connection to the database, and provides links to download the certificates for different AWS regions.

Let's assume that you have downloaded and saved the certificate file `us-east-1-bundle.pem` to `~/rds-certs/` directory on your local machine. 

`pgpx` uses **absolute paths** to load the certificate, so let's assume that the absolute path is the following: `/Users/jon_doe/rds-certs/us-east-1-bundle.pem`.

**3. Run pgpx**

`$ ./pgpx --dbCertPath=/Users/jon_doe/rds-certs/us-east-1-bundle.pem --credsCmd=${DYNAMIC_PASSWORD} --pgHostPort=localhost:5432`

where
- `pgHostPort` is the combination of a host and port of your PostgreSQL server (i.e. same as `${RDS_HOSTNAME}:${RDS_PORT}`)

Once started, `pgpx` will show the welcome message:
```json
{"level":"trace","time":"2023-10-04T17:06:28-04:00","message":"Starting PostgreSQL proxy on localhost:6432 ..."}
```

**4. Use client application to connect to PostgreSQL server**

In this example, we will use [Postico](https://eggerapps.at/postico2/) to connect to PostgreSQL server:

![Postico DB Connection Configuration](postico-connection-configuration.png)

- Put the value of `IAM_USERNAME` in the `User` field
- Put your database name in the `Database` field
- Put the `pgpx` proxy hostname in the 'Host' field (default: `localhost`)
- Put the `pgpx` proxy port in the `Port` field (default: `6432`)
- Put **ANY PASSWORD** in the `Password` field (it will be substituted with the dynamically generated password when database connection is established)

Click "Connect", and enjoy using the database connection via RDS IAM authentication.

## Credits

This work is derived from:

* the amazing [pgproxy](https://github.com/encoredev/encore/tree/main/pkg/pgproxy) - a PostgreSQL proxy implementation package by [Encore](https://encore.dev/)
* the awesome [pgconn](https://github.com/jackc/pgconn) - a PostgreSQL protocol implementation by [Jack Christensen](https://github.com/jackc)



CHANGELOG.md
b792c375 feat: add ignoreUsername as an additional parameter to make decision on overriding credentials (#18)


pgpx.go
package main

import (
	"context"
	"crypto/tls"
	"crypto/x509"
	"flag"
	"fmt"
	"io/ioutil"
	"net"
	"os"
	"os/exec"
	"strings"

	"github.com/rs/zerolog/log"

	"github.com/kurianoff/pgpx/pgproxy"
)

var (
	versionNumber string
)

type tlsLogicalConn struct {
	net.Conn
}

func (c *tlsLogicalConn) Cancel(data *pgproxy.CancelData) error {
	// Implement the cancel logic if necessary or return nil
	// For this example, we'll just return nil since we're not handling cancel requests
	return nil
}

func loadTLSConfig(certFile, keyFile string) (*tls.Config, error) {
	cert, err := tls.LoadX509KeyPair(certFile, keyFile)
	if err != nil {
		return nil, err
	}

	config := &tls.Config{
		Certificates: []tls.Certificate{cert},
		// IMPORTANT: In production, set InsecureSkipVerify to false and setup proper CA verification
		InsecureSkipVerify: true,
	}

	return config, nil
}

func fetchDynamicPassword(cmdStr string) (string, error) {
	cmd := exec.Command("bash", "-c", cmdStr)

	// Replace with your shell command to fetch the dynamic password
	output, err := cmd.CombinedOutput()
	if err != nil {
		log.Error().Msgf("Error executing command: %+v", err)
		return "", err
	}
	return strings.TrimSpace(string(output)), nil
}

func dialBackendWithDynamicPassword(ctx context.Context, startupData *pgproxy.StartupData, pgHostPort string, credsCmd string, credsOverride bool, ignoreUsername bool) (pgproxy.LogicalConn, error) {
	// if ignoreUsername flag not set: find matching username in credsCmd, otherwise do not override password
	overrideCredentials := credsOverride && (ignoreUsername || strings.Contains(strings.ToLower(credsCmd), strings.ToLower(startupData.Username)))
	log.Trace().Msgf("Override credentials: %t", overrideCredentials)

	if overrideCredentials {
		dynamicPassword, err := fetchDynamicPassword(credsCmd)
		if err != nil {
			log.Error().Msgf("Error retrieving dynamic password: %+v", err)
			return nil, err
		}
		startupData.Password = dynamicPassword
	}

	conn, err := net.Dial("tcp", pgHostPort)
	if err != nil {
		log.Error().Msgf("Error connecting to the PostgreSQL backend: %+v", err)
		return nil, err
	}

	log.Trace().Msg("Connection with the PostgreSQL backend has been established.")
	return &tlsLogicalConn{conn}, nil
}

func main() {
	proxyHost := flag.String("proxyHost", "localhost", "Host where pgpx is accepting connections")
	proxyPort := flag.String("proxyPort", "6432", "Port where pgpx is accepting connections")
	pgHostPort := flag.String("pgHostPort", "", "Hostname and port of the PostgreSQL server, e.g. \"localhost:5432\" (required)")
	dbCertPath := flag.String("dbCertPath", "", "Absolute path to the database SSL CA certificate (required)")
	credsCmd := flag.String("credsCmd", "", "Bash command to retrieve temporary database connection password")
	credsOverride := flag.Bool("credsOverride", true, "Indicates whether to override the password provided through the client application")
	ignoreUsername := flag.Bool("ignoreUsername", false, "By default, the tool matches provided username with the one in credsCmd command to decide on overriding the password. Set this flag to ignore username matching.")
	version := flag.Bool("version", false, "Show pgpx version information")

	// Parse the command-line arguments
	flag.Parse()

	if *version == true {
		fmt.Println(versionNumber)
		return
	}

	if (*credsOverride && *credsCmd == "") || *pgHostPort == "" || *dbCertPath == "" {
		fmt.Printf("Usage of %s:\n", os.Args[0])
		flag.PrintDefaults()
		return
	}

	proxyHostPort := fmt.Sprintf("%s:%s", *proxyHost, *proxyPort)

	// Uncomment the following lines if you want the client app to use TLS connection
	//
	// tlsConfig, err := loadTLSConfig("path_to_cert.crt", "path_to_key.key")
	// if err != nil {
	// 	log.Fatalf("Failed to load TLS config: %v", err)
	// }

	ln, err := net.Listen("tcp", proxyHostPort)
	if err != nil {
		log.Error().Msgf("Failed to start listener: %+v", err)
	}

	rootCertPool := x509.NewCertPool()
	pem, err := ioutil.ReadFile(*dbCertPath)
	if err != nil {
		log.Error().Msgf("Failed to load root CA certificate: %+v", err)
		return
	}
	if !rootCertPool.AppendCertsFromPEM(pem) {
		log.Error().Msgf("No certs found in root CA certificate")
		return
	}

	tlsConfig := &tls.Config{
		InsecureSkipVerify: true,
		// If you're using a self-signed certificate or a certificate from a CA that's not
		// in the system's root certificate store, you'll need to add it here.
		RootCAs: rootCertPool,

		// Uncomment and set these if client-side certificate authentication is required by PostgreSQL.
		// Certificates: []tls.Certificate{clientCert},
	}

	proxy := &pgproxy.SingleBackendProxy{
		RequirePassword: true,
		FrontendTLS:     nil,
		BackendTLS:      tlsConfig,
	}

	proxy.DialBackend = func(ctx context.Context, startupData *pgproxy.StartupData) (pgproxy.LogicalConn, error) {
		return dialBackendWithDynamicPassword(ctx, startupData, *pgHostPort, *credsCmd, *credsOverride, *ignoreUsername)
	}

	log.Trace().Msgf("Starting PostgreSQL proxy on %s ...", proxyHostPort)
	err = proxy.Serve(context.Background(), ln)
	if err != nil {
		log.Error().Msgf("Failed to start proxy: %+v", err)
	}
}
